src/Microsoft.VisualStudio.Editor.Razor/DefaultEditorSettingsManager.cs

```diff
@@ -12,26 +13,95 @@ namespace Microsoft.VisualStudio.Editor.Razor
         {
-            var razorLanguageServices = workspaceAccessor.Workspace.Services.GetLanguageServices(RazorLanguage.Name);
-            _editorSettingsManager = razorLanguageServices.GetRequiredService<EditorSettingsManagerInternal>();
+            _foregroundDispatcher = foregroundDispatcher;
+            _settings = EditorSettings.Default;
+            _changedHandlers = new List<WeakReference<EventHandler<EditorSettingsChangedEventArgs>>>();
         }
 
         public override event EventHandler<EditorSettingsChangedEventArgs> Changed
         {
-            add => _editorSettingsManager.Changed += value;
-            remove => _editorSettingsManager.Changed -= value;
+            add
+            {
+                _foregroundDispatcher.AssertForegroundThread();
+
+                _changedHandlers.Add(new WeakReference<EventHandler<EditorSettingsChangedEventArgs>>(value));
+            }
+            remove
+            {
+                _foregroundDispatcher.AssertForegroundThread();
+
+                for (var i = 0; i < _changedHandlers.Count; i++)
+                {
+                    if (_changedHandlers[i].TryGetTarget(out var target) && target == value)
+                    {
+                        _changedHandlers.RemoveAt(i);
+                        return;
+                    }
+                }
+            }
         }
 
-        public override EditorSettings Current => _editorSettingsManager.Current;
+        public override EditorSettings Current
+        {
+            get
+            {
+                lock (SettingsAccessorLock)
+                {
+                    return _settings;
+                }
+            }
+        }
 
-        public override void Update(EditorSettings updateSettings)
+        public override void Update(EditorSettings updatedSettings)
         {
-            _editorSettingsManager.Update(updateSettings);
+            if (updatedSettings == null)
+            {
+                throw new ArgumentNullException(nameof(updatedSettings));
+            }
+
+            _foregroundDispatcher.AssertForegroundThread();
+
+            lock (SettingsAccessorLock)
+            {
+                if (!_settings.Equals(updatedSettings))
+                {
+                    _settings = updatedSettings;
+                    OnChanged();
+                }
+            }
+        }
+
+        private void OnChanged()
+        {
+            _foregroundDispatcher.AssertForegroundThread();
+
+            var handlersToInvoke = new List<EventHandler<EditorSettingsChangedEventArgs>>();
+            for (var i = 0; i < _changedHandlers.Count; i++)
+            {
+                if (_changedHandlers[i].TryGetTarget(out var handler))
+                {
+                    handlersToInvoke.Add(handler);
+                }
+                else
+                {
+                    // Handler was dereferenced, clean it up.
+                    _changedHandlers.RemoveAt(i--);
```

Comment:
There it folks... The Taylor.

Reply:
/cc @Eilon

Reply:
Purposefully using #TheTaylor here... Try writing it backwards you have to insert(0) to maintain handler invocation order, at which point that code needs a comment about un-reversing the handler order adding unnecessary complexity. In this case I actually undid the reverse loop, this is far clearer no matter how much you dislike i--

Reply:
You could just skip the temporary `handlersToInvoke` here, and invoke handlers while doing #TheTaylor then.

Reply:
> You could just skip the temporary handlersToInvoke here, and invoke handlers while doing #TheTaylor then.

Want to capture the handlers prior to invoking them one by one because it's possible for them to remove themselves/add themselves when they're being invoked. That would break the loop entirely.

Reply:
> if (!_changedHandlers[i].TryGetTarget(out var target))
>     {
>         _changedHandlers.RemoveAt(i);
>         continue;
>     }

lol you really don't like i--

Reply:
Ah I see, that's valuable then. My two small reasonable requests:
1. Document in a comment why it's complicated. You just gave me a good reason.
2. Use a while loop when you're going to manipulate `i`

Reply:
Will do